uproot/rootio.py

```diff
@@ -1295,6 +1302,10 @@ def _recarray_dtype(cls, cntvers=False, tobject=True):
 
         return numpy.dtype(dtypesout)
 
+    @classmethod
+    def classname(cls):
+        return cls.__name__
```

Comment:
I'm pretty sure this one is supposed to be

```python
    @property
    def classname(self):
        return self._classname._classname.decode('ascii')
```

Yes: [confirmed here](https://github.com/scikit-hep/uproot/blob/98b9758da71981ecccbe5c0ce040022331b4d7c2/uproot/rootio.py#L869-L872). If you return `cls.__name__`, then that would be the Python name of the class, which would disagree with the classnames defined above in cases with `<`, `>`, `:` characters. With these three `classname`/`classnames`, the user will get C++ names as `str` (with all `<`, `>`, `:` from templates, _not_ translated into underscores), which is what they expect.

Reply:
This is what I meant in the issue when I said I wasn't sure if it would work with complicated class names. I don't think I've ever encountered a `TKey` associated with a class with those characters in the name, but sure, it would be nice to get it to work in that case as well.

The problem here is having it as a `@property`. That means you can't access it from the class without an instance, which is something that I want to be able to do.

Reply:
Right, that's why I dug into it, to at least check the code and try to do the right thing. I've encountered class names with `<`, `>`, `::` in bug reports, but I don't know if I have an example. (I'd like to just try to do the right thing now and if it's broken when we encounter an example, fix it then.)

It _is_ possible to access a property from a class object if that property does not depend on `self`:

```python
>>> class Something:
...     @property
...     def whatever(self):
...         return 999
... 
>>> Something().whatever
999
>>> Something.whatever.fget(None)
999
```

But this one does.

However, I don't see how you could ever get access to one of these classes without having an instance—they're generated from the streamers in the ROOT file: they're only available in `file._context.classes`, which I don't think you're digging into, right?

Reply:
It's quite easy to access the class before you see the actual instance. The way #376 arose was that I was trying to do something like the following to get the class names of objects in a file:

```python
>>> import uproot
>>> file = uproot.open('055400.root')
>>> file.classes()[0][1]._classname
b'TTree'
>>> file.classes()[1][1]._classname
<property object at 0x7f7aeeb8fdd0>
```

As you can see, the TTree produces a `bytes` object and the second (a TNamed) returns a property, which is quite odd and unhelpful. It's true that I could instantiate the TNamed to get `_classname`, but that seems very clunky, especially since the behavior is different from the TTree and the fact that the class name is intrinsic to the class and not a true instance property.

Reply:
Oh, darn, you're right. In that case, the way this should be done is to burn in another attribute on the class, at https://github.com/scikit-hep/uproot/blob/98b9758da71981ecccbe5c0ce040022331b4d7c2/uproot/rootio.py#L869-L872

```python
            if sys.version_info[0] > 2:
                code.insert(0, "    _classname = {0}".format(repr(streamerinfo._fName)))
            else:
                code.insert(0, "    _classname = b{0}".format(repr(streamerinfo._fName)))
            code.insert(0, "    classname = {0}".format(repr(streamerinfo._fName.decode("ascii")))
```

The `streamerinfo._fName` is the original C++ class name; we just need to decode that into a `str` and add it to the auto-generated code for the class.

Reply:
Right, that's what I started to do originally with this, but the problem is you would then have to deal with `builtin_classes` individually. That's because the code you pointed to only runs under this condition:

https://github.com/scikit-hep/uproot/blob/eb5d8862f8a382606be7bb9fa84f169b7cd19513/uproot/rootio.py#L676

For example, what you wrote works for TTree but not for TNamed. I'd have to go into the definition of each of these:

https://github.com/scikit-hep/uproot/blob/eb5d8862f8a382606be7bb9fa84f169b7cd19513/uproot/rootio.py#L1520-L1538

and add a `classname = ` line. This is why I instead tried to write a `classmethod`.

Reply:
I decided to go with slightly WETter code than have this inconsistency. See 611308f81cb8e79f96e2b4359fa590600d414bf8 .